statzone.1
.\"
.\" Copyright (c) 2012-2023, Frederic Cambus
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are met:
.\"
.\"   * Redistributions of source code must retain the above copyright
.\"     notice, this list of conditions and the following disclaimer.
.\"
.\"   * Redistributions in binary form must reproduce the above copyright
.\"     notice, this list of conditions and the following disclaimer in the
.\"     documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd $Mdocdate: March 13 2023 $
.Dt STATZONE 1
.Os
.Sh NAME
.Nm statzone
.Nd DNS zone file analyzer targeted at TLD zones
.Sh SYNOPSIS
.Nm
.Op Fl hv
.Ar zonefile
.Sh DESCRIPTION
.Nm
is a DNS zone file analyzer targeted at TLD zones.
.Pp
After analyzing a zone, it returns counts for: IPv4 and IPv6 glue records,
NS records (total and uniques), DS records, DNSSEC signed domains, IDNs
domains, and total number of domains.
.Pp
It is currently used to generate TLD Zone File Statistics on StatDNS.
.Pp
If
.Ar file
is a single dash (`-'),
.Nm
reads from the standard input.
.Pp
The options are as follows:
.Bl -tag -width 10n
.It Fl h
Display usage.
.It Fl v
Display version.
.El
.Sh AUTHORS
.Nm
was written by
.An Frederic Cambus .


ChangeLog
StatZone 1.1.1 (2021-11-16)

- Switch to using nullptr instead of NULL
- Fix erroneous comment about timer value in summary()
- Remove code to get zone file size, it's unused
- Update the arpa.zone copy used for testing
- Add SPDX short license identifier in source file
- Use CMAKE_CXX_STANDARD instead of hardcoding compiler flags
- Remove dependency on uthash, it is not used anymore
- Switch to using getline(3) instead of fgets(3)



StatZone 1.1.0 (2021-04-01)

- Remove "USAGE:" prefix when displaying usage information
- Add missing periods when displaying options in usage(), sync README
- StatZone is now a C++ project and requires a C++11 compiler
- Use std::unordered_set instead of uthash to count unique strings
- Remove now unneeded dependency on uthash
- Use std::chrono to calculate elapsed time in summary()
- Remove now unneeded timespecsub macro in compat header
- Remove now unneeded CLOCK_MONOTONIC fallback macro in compat header
- Use iostream to output CSV values
- Start using std::string for domain and previous_domain
- Remove now unused error() function
- Use std::basic_string::compare instead of strncmp(3)
- Add a signal handler for SIGINFO, wrapping summary()
- Remove pledge() support and get rid of the compat directory entirely
- Bump CMake minimum required version to 2.8.12



StatZone 1.0.5 (2021-02-08)

- Stop forcing FORTIFY_SOURCE level 2, it should be package builders decision
- Remove hardcoded program name in the manual page
- Rename displayUsage() and displaySummary() to usage() and summary()
- Get rid of most global variables, move all declarations to main()
- Remove dead increments for argc and argv
- Refactor things to check earlier on if *lineBuffer is not NUL
- Refactor the domain comparison checks, allows removing strdup("") call
- Add an error() function to wrap errx() calls
- Check malloc() and strdup() return value and error out on failed allocations
- Group and initialize uthash struct declarations
- Free previous_domain to appease Clang's static analyzer



StatZone 1.0.4 (2020-10-19)

- Add support for seccomp on arm
- Update error message to use seccomp architecture names
- Improve experimental seccomp support mention in the README
- Make displayUsage() and displaySummary() static
- Set CMAKE_BUILD_TYPE to "Release" in CMakeLists.txt
- Enable FORTIFY_SOURCE level 2



StatZone 1.0.3 (2020-07-06)

- Remove some unneeded seccomp related includes
- Validate architectures for seccomp
- Add seccomp support on i386, tested on glibc and musl
- Use __NR_ instead of SYS_ prefix in #if defined checks



StatZone 1.0.2 (2020-06-10)

- Use CLOCK_REALTIME if CLOCK_MONOTONIC is not available
- Remove unneeded <sys/select.h> include
- Define _GNU_SOURCE in CMakeLists.txt to avoid cluttering source files
- Allow the openat and write syscalls, they are used on glibc systems
- No need to set HAVE_SECCOMP to 1, defining it is enough
- Introduce the displaySummary() function, and use it to print summary
- Move elapsed time calculation in the displaySummary() function
- Add a signal handler for SIGINFO, to display progression on
  supported platforms
- Do not use intermediate variable to calculate elapsed time
- Remove now useless runtime member from the results structure



StatZone 1.0.1 (2020-03-30)

- Add install target for man page
- Add a test case, using the arpa.zone
- Improve program description in README and in the man page
- Lint the manual page, add missing paragraph break macros
- Mention that seccomp is currently only optional



StatZone 1.0.0 (2020-03-22)

- Initial release


src/statzone.c
/*
 * StatZone 1.0.5
 * Copyright (c) 2012-2021, Frederic Cambus
 * https://www.statdns.com
 *
 * Created: 2012-02-13
 * Last Updated: 2021-02-12
 *
 * StatZone is released under the BSD 2-Clause license
 * See LICENSE file for details.
 */

#include <err.h>
#include <getopt.h>
#include <inttypes.h>
#include <signal.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>
#include <time.h>

#ifdef HAVE_SECCOMP
#include <sys/prctl.h>
#include <linux/seccomp.h>
#include "seccomp.h"
#endif

#include <uthash.h>

#include "compat.h"
#include "config.h"
#include "strtolower.h"

struct timespec begin, current, elapsed;
struct results results;

static void
error(const char *str)
{
	errx(EXIT_FAILURE, "%s", str);
}

static void
usage()
{
	printf("statzone [-hv] zonefile\n\n" \
	    "The options are as follows:\n\n" \
	    "	-h	Display usage\n" \
	    "	-v	Display version\n");
}

static void
summary()
{
	/* Stopping timer */
	clock_gettime(CLOCK_MONOTONIC, &current);
	timespecsub(&current, &begin, &elapsed);

	/* Print summary */
	fprintf(stderr, "Processed %" PRIu64 " lines in %f seconds.\n",
	    results.processedLines,
	    elapsed.tv_sec + elapsed.tv_nsec / 1E9);
}

int
main(int argc, char *argv[])
{
	struct stat zonefile_stat;

	struct domain {
		char *domain;
		UT_hash_handle hh;
	};

	struct domain *ds = NULL, *signed_domains = NULL;
	struct domain *ns = NULL, *unique_ns = NULL;

	int opt, token_count;

	char linebuffer[LINE_LENGTH_MAX];
	char *input, *domain, *previous_domain = NULL;
	char *rdata, *token = NULL, *token_lc = NULL;

	FILE *zonefile;

	if (pledge("stdio rpath", NULL) == -1) {
		err(EXIT_FAILURE, "pledge");
	}

#ifdef HAVE_SECCOMP
	if (prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0)) {
		perror("Can't initialize seccomp");
		return EXIT_FAILURE;
	}

	if (prctl(PR_SET_SECCOMP, SECCOMP_MODE_FILTER, &statzone)) {
		perror("Can't load seccomp filter");
		return EXIT_FAILURE;
	}
#endif

#ifdef SIGINFO
	signal(SIGINFO, summary);
#endif

	while ((opt = getopt(argc, argv, "hv")) != -1) {
		switch (opt) {

		case 'h':
			usage();
			return EXIT_SUCCESS;

		case 'v':
			printf("%s\n", VERSION);
			return EXIT_SUCCESS;
		}
	}

	if (optind < argc) {
		input = argv[optind];
	} else {
		usage();
		return EXIT_SUCCESS;
	}

	/* Starting timer */
	clock_gettime(CLOCK_MONOTONIC, &begin);

	/* Open zone file */
	if (!strcmp(input, "-")) {
		/* Read from standard input */
		zonefile = stdin;
	} else {
		/* Attempt to read from file */
		if (!(zonefile = fopen(input, "r"))) {
			perror("Can't open zone file");
			return EXIT_FAILURE;
		}
	}

	/* Get zone file size */
	if (fstat(fileno(zonefile), &zonefile_stat)) {
		perror("Can't stat zone file");
		return EXIT_FAILURE;
	}

	while (fgets(linebuffer, LINE_LENGTH_MAX, zonefile)) {
		if (!*linebuffer)
			continue;

		if (*linebuffer == ';') /* Comments */
			continue;

		if (*linebuffer == '$') /* Directives */
			continue;

		token_count = 0;
		token = strtok(linebuffer, " \t");

		if (token)
			domain = strtolower(token);

		while (token) {
			if (*token == ';') { /* Comments */
				token = NULL;
				continue;
			}

			token_lc = strtolower(token);
			if (token_count && !strcmp(token_lc, "nsec")) {
				token = NULL;
				continue;
			}

			if (token_count && !strcmp(token_lc, "nsec3")) {
				token = NULL;
				continue;
			}

			if (token_count && !strcmp(token_lc, "rrsig")) {
				token = NULL;
				continue;
			}

			if (token_count && !strcmp(token_lc, "a"))
				results.a++;

			if (token_count && !strcmp(token_lc, "aaaa"))
				results.aaaa++;

			if (token_count && !strcmp(token_lc, "ds")) {
				results.ds++;

				HASH_FIND_STR(signed_domains, domain, ds);

				if (!ds) {
					ds = malloc(sizeof(struct domain));
					if (ds == NULL)
						error("Memory allocation error.");

					ds->domain = strdup(domain);
					if (ds->domain == NULL)
						error("Memory allocation error.");

					HASH_ADD_STR(signed_domains, domain, ds);
				}
			}

			if (!strcmp(token_lc, "ns")) {
				results.ns++;

				if (previous_domain == NULL ||
				    strlen(previous_domain) != strlen(domain) ||
				    strncmp(domain, previous_domain, strlen(domain))) {
					results.domains++;

					free(previous_domain);
					previous_domain = strdup(domain);

					if (previous_domain == NULL)
						error("Memory allocation error.");

					if (!strncmp(domain, "xn--", 4))
						results.idn++;
				}

				rdata = strtok(NULL, "\n");

				if (rdata && strchr(rdata, ' '))
					rdata = strtok(NULL, "\n");

				if (rdata) {
					HASH_FIND_STR(unique_ns, rdata, ns);

					if (!ns) {
						ns = malloc(sizeof(struct domain));
						if (ns == NULL)
							error("Memory allocation error.");

						ns->domain = strdup(rdata);
						if (ns->domain == NULL)
							error("Memory allocation error.");

						HASH_ADD_STR(unique_ns, domain, ns);
					}
				}
			}

			token = strtok(NULL, " \t");
			token_count++;
		}

		results.processedLines++;
	}

	/* Don't count origin */
	if (results.domains)
		results.domains--;

	/* Printing CVS values */
	fprintf(stdout, "---[ CSV values ]--------------------------------------------------------------\n");
	fprintf(stdout, "IPv4 Glue,IPv6 Glue,NS,Unique NS,DS,Signed,IDNs,Domains\n");
	fprintf(stdout, "%" PRIu64 ",", results.a);
	fprintf(stdout, "%" PRIu64 ",", results.aaaa);
	fprintf(stdout, "%" PRIu64 ",", results.ns);
	fprintf(stdout, "%u,", HASH_COUNT(unique_ns));
	fprintf(stdout, "%" PRIu64 ",", results.ds);
	fprintf(stdout, "%u,", HASH_COUNT(signed_domains));
	fprintf(stdout, "%" PRIu64 ",", results.idn);
	fprintf(stdout, "%" PRIu64 "\n", results.domains);

	/* Printing results */
	summary();

	/* Clean up */
	free(previous_domain);
	fclose(zonefile);

	return EXIT_SUCCESS;
}
